README.md
# @bigtest/convergence [![CircleCI](https://circleci.com/gh/bigtestjs/convergence/tree/master.svg?style=svg)](https://circleci.com/gh/bigtestjs/convergence/tree/master)

Recognize a desired state and synchronize on when that state has been achieved.

## Why Convergence?

Let's say you want to write an assertion to verify a simple cause and
effect: when a certain button is clicked, a dialog appears containing
some text that gets loaded from the network.

In order to do this, you have to make sure that your assertion runs
_after_ the effect you're testing has been realized.

![Image of assertion after an effect](https://raw.githubusercontent.com/thefrontside/bigtest/master/packages/convergence/images/assertion-after.png)

If not, then you could end up with a false negative, or "flaky test"
because you ran the assertion too early. If you'd only waited a little
bit longer, then your test would have passed. So sad!

![Image of false negative test](https://raw.githubusercontent.com/thefrontside/bigtest/master/packages/convergence/images/false-negative.png)

In fact, test flakiness is the reason most people shy away from
writing big tests in JavaScript in the first place. It seems almost
impossible to write robust tests without having visibility into the
internals of your runtime so that you can manually synchronize on
things like rendering and data loading. Unfortunately, those can be a
moving target, and worse, they couple you to your framework.

But what if instead of trying to run our assertions at just the right
time, we ran them _many_ times until they either pass or we decide to
give up?

![Image of convergent assertion](https://raw.githubusercontent.com/thefrontside/bigtest/master/packages/convergence/images/convergent-assertion.png)

This is the essence of what `@bigtest/convergence` provides:
repeatedly testing for a condition and then allowing code to run when
that condition has been met.

And it isn't just for assertions either. Because it is a general
mechanism for synchronizing on any observed state, It can be used to
properly time test setup and teardown as well.

## Creating a Convergence

A convergent function is a function that runs repeatedly until it no
longer returns false or throws an error. When the function is finally
successfully executed, it is considered to be passing and a converging
promise will resolve. However, if the converging function does not pass
within the provided timeout period the promise will reject with the
last error thrown from the function.

### Using `Convergence`

The `Convergence` class allows you to create a convergence, add
assertions to converge on, and execute the convergent assertions all
within a single timeout period.

``` javascript
import Convergence from '@bigtest/convergence'

// starts a new queue
new Convergence()
  // adds a convergent function to the queue
  .when(() => expect($el).to.exist)
  // called when the previous function converges
  .do(() => $el.get(0).click())
  // adds a convergent function that resolves when it is true
  // for the remaining timeout period
  .always(() => expect($el).to.have.prop('disabled', true))
  // starts converging and returns a promise that resolves
  // when all convergences have been met
  .run()
```

#### Immutability

`Convergence` has an immutable interface, so it's methods will
return new instances.

``` javascript
// will converge when the total equals 5
let convergeWhen = new Convergence()
  .when(() => total === 5);

// will log the total after the first convergence
let convergeAndLog = convergeWhen
  .do(() => console.log(total));

// after logging the total, will converge when it remains 5 for the
// duration of the remaining timeout
let convergeAlways = convergeAndLog
  .always(() => total === 5);

// all three convergences can be ran in parallel
convergeWhen.run();
convergeAndLog.run();
convergeAlways.run();
```

_Note: `.run()` does not return a new instance, instead it returns a promise._

---

#### Syntax

``` javascript
new Convergence([timeout = 2000]);
```

#### Parameters

**`timeout`**

You can define the convergences total timeout during initialization of
a `Convergence` instance. This timeout will be used for the set of
convergent functions unless the `.timeout()` method is used to change
the total timeout period.

``` javascript
let converge = new Convergence(1000);
```

#### Methods

**`.timeout([timeout])`**

Given a timeout (in `ms`), the `.timeout()` method will return a new
`Convergence` instance with the new timeout period replacing the old
one. When no argument is given, this method will return the current
timeout period of the convergence.

``` javascript
let convergeLong = converge.timeout(5000);

converge.timeout();  // => 1000
convergeLong.timeout(); // => 5000
```

**`.when(assert)`**

Returns a new `Convergence` instance and adds the provided assertion
to its queue. When this instance is ran, the `assert` function will be
looped over repeatedly until it passes, or until the convergence's
timeout has been exceeded.

``` javascript
// this convergence will resolve when `total` equals `5`
converge.when(() => total === 5)
```

**`.always(assert[, timeout])`**

Another common pattern is asserting that something **has not**
changed. With a typical convergence, the state may change after an
assertion converges immediately. So for these scenarios, you want to
converge _when the assertion passes for the duration of a timeout._
`.always()` is just like `.when()` above, except that the `assert`
function is looped over repeatedly until it fails for the first time
or never fails for the duration of the timeout.

``` javascript
// this convergence will resolve when total is always 5
// for the duration of the remaining timeout
converge.always(() => total === 5)
```

When a convergence added with `.always()` is last in the queue, it
will be given the remaining total timeout to converge on its assertion
always passing. When it _is not_ last in the queue, `timeout` is used
instead. By default, `timeout` will be a tenth of the total timeout
or `20ms` (whichever is greater).

``` javascript
converge
  // will be given any remaining timeout
  .always(() => total === 5)

converge
  // will use the 500ms timeout instead
  .always(() => total === 5, 500)
  .when(() => total === 10)

new Convergence(2000)
  // defaults to 200ms
  .always(() => total === 5)
  // this counts as part of the queue
  .do(() => console.log(total))
```

**`.do(exec)`**

This method is useful when you need to execute something after a
convergence in the queue, but before other convergences are ran.
This can help with debugging in between convergences and also
allows you to run side effects between convergences.

``` javascript
converge
  .when(() => total === 5)
  // executes after the total is equal to 5
  .do(() => total *= 100)
  // starts converging after the total has been multiplied
  .always(() => total === 500)
```

Functions in a `Convergence` queue curry their return value
between other functions in the queue. This means that what you return
from one function in the queue will be available as the argument
given to the next function in the queue.

``` javascript
converge
  // returns the element when it exists in the DOM
  .when(() => {
    let $el = $('[data-test-element]');
    expect($el).to.exist;
    return $el;
  })
  // clicks the element and returns it for chaining
  .do(($el) => {
    $el.get(0).click();
    return $el;
  })
  // converges when the element is disabled for the remaining timeout
  .always(($el) => {
    expect($el).to.have.prop('disabled', true)
  })
```

You can even return promises, or other convergences from this method.

``` javascript
converge
  .do(() => Promise.resolve(5))
  .do((total) => new Convergence()
    // the final always will still get the remaining time
    .always(() => total === 5 && total))
  // ... unless it is followed by additional methods
  .do((total) => total *= 100)
```


**`.append()`**

Combines convergences to allow composing them together to create brand
new convergence instances.

``` javascript
let converge1 = converge.when(() => total === 1)
let converge5 = converge.when(() => total === 5)

// converges when the total first equals `1` and then equals `5`
converge1.append(converge5)
```

**`.run()`**

In order to actually run a `Convergence` instance, you must call the
`.run()` method. This method **does not return another instance**,
instead it returns a promise that resolves when all assertions in the
queue have converged. The returned promise will keep track of the
current timeout and ensure that all convergences in the queue converge
within that period.

Because `.run()` returns a promise, and `Convergence` is immutable,
you can call `.run()` multiple times on the same convergence instance.

``` javascript
converge.run()
  .catch(() => {
    console.log('failed. trying again...');
    return converge.run();
  })
  .then(() => {
    console.log('success!')
  });
```

The promise returned from `.run()` resolves with a stats object. This
stats object holds various information about how the convergences in
the queue ran. The `queue` property is an array of stats objects
specific to each function in the queue.

``` javascript
converge.run((stats) => {
  stats.start    // start time of the convergences
  stats.end      // end time of the convergences
  stats.elapsed  // time taken to converge on the entire queue
  stats.runs     // number of times functions in the queue ran
  stats.timeout  // this Convergence instance's timeout
  stats.value    // value returned from the last function in the queue
  stats.queue    // an array of stats objects from the queue

  // each function in the queue produces similar stats objects
  stats.queue[0].start    // start time of this convergence
  stats.queue[0].end      // end time of this convergence
  stats.queue[0].elapsed  // time taken for this convergence
  stats.queue[0].runs     // number of times this convergence ran
  stats.queue[0].timeout  // the timeout this convergence was given
  stats.queue[0].always   // whether this convergence used .always()
  stats.queue[0].value    // value returned from this convergence
});
```

_Note: functions added using `.do()` produce a small stats object as
well, but they omit `runs`, `timeout`, and `always` since they do not
apply to these functions._

### Helpers

In addition to the `Convergence` class, this package also exports a
few convergence helpers.

**`isConvergence(object)`**

Returns `true` if the given `object` implements a `Convergence`
interface. A `Convergence` interface consists of a queue, a
`#timeout()` method, and a `#run()` method.

``` javascript
class CustomConvergence extends Convergence() {
  ...
}

isConvergence(new Convergence()) // => true
isConvergence(new CustomConvergence()) // => true
isConvergence(new Array()) // => false
```

**`when(assertion[, timeout=2000])`**

Starts converging on the given `assertion`, resolving when it passes
_within_ the timeout period. The assertion will run once every 10ms
and is considered to be passing when it does not error or return
false. If the assertion never passes within the timeout period, the
promise will reject as soon as it can with the last error it recieved.

``` javascript
// simple boolean test
await when(() => total === 100)

// with chai assertions
await when(() => {
  expect(total).to.equal(100)
  expect(add(total, 1)).to.equal(101)
})
```

Returns a thennable function that can both be used as a callback and
awaited on directly. For example, testing frameworks that support
async tests work well with convergent assertions.

``` javascript
test('will eventually become bar within 1s', when(() => {
  expect(foo).to.equal('bar');
}, 1000));
```

_Note: `when` will throw an error **after** it fails to converge
within the timeout. When using with testing frameworks, be aware of
any test timeouts that may occur before the convergence fails._

**`always(assertion[, timeout=200])`**

Starts converging on the given `assertion`, resolving when it passes
_throughout_ the timeout period. Like `when()`, The assertion will run
once every 10ms and is considered to be passing when it does not error
or return false. However, if the assertion does not pass consistently
throughout the entire timeout period, it will reject the very first
time it encounters a failure.

```javascript
// simple boolean test
await always(() => total !== 100)

// with chai assertions
await always(() => {
  expect(total).to.not.equal(100)
  expect(add(total, 1)).to.equal(101)
})
```

Also returns a thennable function that can both be used as a callback
and awaited on directly.

``` javascript
test('remains foo for at least 1s', always(() => {
  expect(foo).to.equal('foo');
}, 1000));
```

_Note: `always` will resolve **after** it converges throughout the
entire timeout. When using with testing frameworks, be aware of any
test timeouts that may occur before the convergence resolves._


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)

## [Unreleased]

### Fixed

- Make sure $root descriptor function is not wrapped in the chaining mechanism 

## [1.1.1] - 2018-10-23

### Fixed

- global Date reference which caused date & time mocking to break
  convergences

## [1.1.0] - 2018-10-04

### Added

- queued functions retain the previous value when returning undefined

### Changed

- trailing `always` assertions on Convergence instances default to
  remaining time instead of always using remaining time

## [1.0.0] - 2018-09-13

### Changed

- immediately throw when encountering an async assertion
- `when` and `always` helpers return thennable functions that can be
  used as callbacks, or awaited on directly
- updated to stable dependencies

### Removed

- deprecated `once` method
- Typescript types

## [0.10.0] - 2018-07-05

### Added

- error when using async functions or returning promises with
  convergent assertions

## [0.9.1] - 2018-05-05

### Changed

- internal docs links
- update `@bigtest/meta`

## [0.9.0] - 2018-05-05

### Changed

- "stack" terminology to "queue" to better reflect first-in-first-out

### Added

- `when` and `always` helpers to immediately start converging on a
single assertion without the need for creating a `Convergence`

### Removed

- `convergeOn` export in favor of `when` and `always`

## [0.8.0] - 2018-04-30

### Added

- transpiled es module bundle
- typescript definitions

### Changed

- `_timeout` to `timeout` when providing an options hash to the
convergence constructor

## [0.7.0] - 2018-04-07

### Added

- `docs` script to generate documentation
- `.when()` to replace the `.once()` method

### Changed

- updated inline api docs
- `Convergence` assertions retain the instance context

### Deprecated

- `.once()` in favor of `.when()`

### Removed

- context currying from `convergeOn`

## [0.6.0] - 2018-03-16

### Added

- then method to support async / await syntax without calling `.run()`

### Fixed

- `.append()` method to use `isConvergence` instead of `instanceof`

## [0.5.0] - 2018-03-14

### Added

- full support for returning promises in `.do()` blocks
- timeout error thrown when convergence exceeds timeout
- support for returning convergence in `.do()` blocks
- `isConvergence` helper function

## [0.4.0] - 2018-03-02

### Changed

- convergent assertions no longer attempt to resolve or reject before
their timeout

## [0.3.0] - 2018-02-18

### Added

- ability to extend the convergence class

### Changed

- convergence constructor to be more general to allow subclasses to
handle their own construction

## [0.2.0] - 2018-02-16

### Added

- `.append()` method to combine convergences together

- "module" entry point to support native consumption of @bigtest/mocha
as es module

### Fixed

- correctly remove comments from compiled files


src/convergence.js
import {
  isConvergence,
  runAssertion,
  runCallback
} from './utils';

/**
 * ```
 * import Convergence from '@bigtest/convergence'
 * ```
 *
 * Convergences are powerful, immutable, reusable, and composable
 * assertions that allow you to know immediately when a desired state
 * is achieved.
 *
 * ``` javascript
 * setTimeout(() => foo = 'bar', 100)
 * await new Convergence().when(() => foo === 'bar')
 * console.log(foo) // => "bar"
 * ```
 *
 * By default, a convergence will converge before or after `2000ms`
 * depending on the type of assertions defined. This can be configured
 * by providing a timeout when initializing the convergence, or by
 * using the [`#timeout()`](#/Convergence#timeout) method.
 *
 * ``` javascript
 * new Convergence(100)
 * new Convergence().timeout(5000)
 * ```
 *
 * Using [`#when()`](#when), the assertions will run multiple times
 * until they pass. Similarly, [`#always()`](#always) ensures that
 * assertions keep passing for a period of time.
 *
 * ``` javascript
 * // converges when `foo` is equal to `'bar'` within 100ms
 * new Convergence(100).when(() => foo === 'bar')
 * // converges after `foo` is equal to `'bar'` for at least 100ms
 * new Convergence(100).always(() => foo === 'bar')
 * ```
 *
 * Convergences are immutable, and as such, it's methods return new
 * instances. This allows you to compose multiple convergences and
 * start each of them separately using their respective
 * [`#run()`](#/Convergence#run) methods.
 *
 * ``` javascript
 * let converge = new Convergence(300)
 * let convergeFoo = converge.when(() => foo === 'foo')
 * let convergeFooBar = convergeFoo.when(() => foo === 'bar')
 * let convergeFooBarBaz = convergeFooBar.when(() => foo === 'baz')
 *
 * setTimeout(() => foo = 'foo', 100)
 * setTimeout(() => foo = 'bar', 200)
 * setTimeout(() => foo = 'baz', 150)
 *
 * // resolves after 100ms
 * convergeFoo.run()
 * // resolves after 200ms
 * convergeFooBar.run()
 * // rejects after 300ms since it wasn't `baz` _after_ `bar`
 * convergeFooBarBaz.run()
 * ```
 *
 * Convergences are also thennable, which immediately invokes
 * [`#run()`](#/Convergence#run). This allows them to be able to be used anywhere
 * Promises can be used in most cases.
 *
 * ``` javascript
 * async function onceBarAlwaysBar() {
 *   await new Convergence()
 *     .when(() => foo === 'bar')
 *     .always(() => foo === 'bar')
 * }
 *
 * Promise.race([
 *   onceBarAlwaysBar(),
 *   new Convergence().when(() => foo === 'baz')
 * ])
 * ```
 */
class Convergence {
  /**
   * The constructor actually takes two params, `options` and
   * `previous`. Publicly, `options` is `timeout`, but internally, new
   * instances receive new `options` in addition to the `previous`
   * instance. This allows things that extend convergences to still be
   * immutable, but requires that they have deeper knowledge of the
   * internal API.
   *
   * @constructor
   * @param {Number} timeout - Initial convergence timeout
   */
  constructor(options = {}, previous = {}) {
    // a timeout was given
    if (typeof options === 'number') {
      options = { timeout: options };
    }

    let {
      timeout = previous._timeout || 2000,
      _queue = []
    } = options;

    // merge with the previous queue, if given
    _queue = [...(previous._queue || []), ..._queue];

    Object.defineProperties(this, {
      _timeout: { value: timeout },
      _queue: { value: _queue }
    });
  }

  /**
   * Returns a new convergence instance with the given timeout,
   * inheriting the current instance's assertions. If no timeout is
   * given, returns the current timeout for this instance.
   *
   * ``` javascript
   * let quick = new Convergence(100)
   * let long = quick.timeout(5000)
   *
   * quick.timeout() // => 100
   * long.timeout() // => 5000
   * ```
   *
   * @param {Number} [timeout] - Timeout for the next convergence
   * @returns {Number|Convergence} The current instance timeout or
   * a new convergence instance
   */
  timeout(timeout) {
    if (typeof timeout !== 'undefined') {
      return new this.constructor(timeout, this);
    } else {
      return this._timeout;
    }
  }

  /**
   * Returns a new convergence instance with an additional assertion.
   * This assertion is run repeatedly until it passes within the
   * timeout. If the assertion does not pass within the timeout, the
   * convergence will fail.
   *
   * ``` javascript
   * // would converge when `foo` equals `'bar'`
   * let convergeFoo = new Convergence().when(() => foo === 'bar')
   *
   * // would converge when `foo` equals `'bar'` and then `'baz'`
   * let convergeFooBar = convergeFoo.when(() => foo === 'baz')
   * ```
   *
   * @param {Function} assertion - The assertion to converge on
   * @returns {Convergence} A new convergence instance
   */
  when(assertion) {
    return new this.constructor({
      _queue: [{ assertion }]
    }, this);
  }

  /**
   * Returns a new convergence instance with an additional assertion.
   * This assertion is run repeatedly to ensure it passes throughout
   * the timeout. If the assertion fails at any point during the
   * timeout, the convergence will fail.
   *
   * ``` javascript
   * // would converge after `foo` remains `'foo'` for at least 100ms
   * new Convergence(100).always(() => foo === 'foo')
   * ```
   *
   * When an always assertion is encountered at the end of a
   * convergence, the timeout defaults to the remaining time for the
   * current running instance; minumum `20ms`. When not at the ned of
   * a convergence, it defaults to one-tenth of the total timeout.
   *
   * ``` javascript
   * let convergeFooThenBar = new Convergence(1000)
   * // would continue after `foo` remains `'foo'` for at least 100ms
   *   .always(() => foo === 'foo')
   * // then have any time remaining to converge on `foo` being `'bar'`
   *   .when(() => foo === 'bar')
   * ```
   *
   * Given a timeout, it is capped at the remaining timeout for the
   * current running instance.
   *
   * ``` javascript
   * let convergeFooThenBar = new Convergence(100)
   * // would continue after `foo` remains `'foo'` for at least 50ms
   *   .always(() => foo === 'foo', 50)
   * // then have 50ms remaining to converge on `foo` being `'bar'`
   *   .when(() => foo === 'bar')
   * // and a maximum of ~50ms to converge on it remaining `bar`
   *   .always(() => foo === 'bar', 100)
   * ```
   *
   * @param {Function} assertion - The assertion to converge on
   * @param {Number} [timeout] - The timeout to use, capped at the
   * remaining timeout.
   * @returns {Convergence} A new convergence instance
   */
  always(assertion, timeout) {
    return new this.constructor({
      _queue: [{
        always: true,
        assertion,
        timeout
      }]
    }, this);
  }

  /**
   * Returns a new convergence instance with a callback added to its
   * queue. When a running convergence instance encounters a callback,
   * it will be invoked with the value returned from the last function
   * in the queue. The resulting return value will also be provided to
   * the following function in the queue. If the return value is
   * undefined, the previous return value will be retained.
   *
   * ``` javascript
   * new Convergence()
   *   // continues after finding a random even number
   *   .when(() => {
   *     let n = Math.ceil(Math.random() * 100)
   *     return !(n % 2) && n
   *   })
   *   // multiplies the even number by another random number
   *   .do((even) => {
   *     return even * Math.ceil(Math.random() * 100)
   *   })
   *   // does not return, the previous value will be retained
   *   .do((rand) => {
   *     console.log('even number times random number = ', rand);
   *   })
   *   // asserts that any number times an even number is even
   *   .always((rand) => {
   *     let rand = Math.ceil(Math.random() * 100);
   *     return !(rand % 2)
   *   }, 100)
   * ```
   *
   * When a promise is returned from a callback, the convergence will
   * wait for the promise to resolve before continuing.
   *
   * ``` javascript
   * new Convergence()
   *   .when(() => foo === 'bar')
   *   .do(() => doSomethingAsync())
   *   .do((baz) => console.log('resolved with', baz))
   * ```
   *
   * Returning other convergences from a callback is also
   * supported. The returned convergence will be run with the current
   * remaining timeout. This is useful when computing convergences
   * after converging on another state.
   *
   * ``` javascript
   * new Convergence()
   *   // continue when `num` is less than 100
   *   .when(() => num < 100)
   *   .do(() => {
   *     // if odd, wait until it is even
   *     if (num % 2) {
   *       return new Convergence()
   *         .when(() => !(num % 2) && num)
   *     } else {
   *       return num;
   *     }
   *   })
   * ```
   *
   * @param {Function} callback - The callback to execute
   * @returns {Convergence} A new convergence instance
   */
  do(callback) {
    return new this.constructor({
      _queue: [{ callback }]
    }, this);
  }

  /**
   * Appends another convergence's queue to this convergence's queue
   * to allow composing different convergences together.
   *
   * ``` javascript
   * // would converge when `foo` equals `'bar'`
   * let convergeBar = new Convergence().when(() => foo === 'bar')
   *
   * // would converge when `foo` equals `'baz'`
   * let convergeBaz = new Convergence().when(() => foo === 'baz')
   *
   * // would converge when `foo` equals `'bar'` and then `'baz'`
   * let convergeBarBaz = convergeBar.append(convergeBaz)
   * ```
   *
   * @param {Convergence} convergence - A convergence instance
   * @returns {Convergence} A new convergence instance
   */
  append(convergence) {
    if (!isConvergence(convergence)) {
      throw new Error('.append() only works with convergence instances');
    }

    return new this.constructor({
      _queue: convergence._queue
    }, this);
  }

  /**
   * Runs the current convergence instance, returning a promise that
   * will resolve after all assertions have converged, or reject when
   * any of them fail.
   *
   * ``` javascript
   * let convergence = new Convergence().when(() => foo === 'bar');
   *
   * // will converge within the timeout or fail afterwards
   * convegence.timeout(100).run()
   *   .then(() => console.log('foo is bar!'))
   *   .catch(() => console.log('foo is not bar'))
   * ```
   *
   * When an assertion fails and the convergence rejects, it is
   * rejected with the last error thrown from the assertion.
   *
   * ``` javascript
   * let convergence = new Convergence().when(() => {
   *   expect(foo).to.equal('bar')
   * })
   *
   * // will fail after 100ms if `foo` does not equal `'bar'`
   * convegence.timeout(100).run()
   *   .catch((e) => console.error(e)) // expected '' to equal 'bar'
   * ```
   *
   * When the convergence is successful and the promise resolves, it
   * will resolve with a stats object containing useful information
   * about how the convergence and it's assertions ran.
   *
   * ``` javascript
   * let convergence = new Convergence()
   *   .when(() => foo === 'bar')
   *   .always(() => foo === 'bar')
   *
   * convergence.run().then((stats) => {
   *   stats.start   // timestamp of the convergence start time
   *   stats.end     // timestamp of the convergence end time
   *   stats.elapsed // amount of milliseconds the convergence took
   *   stats.timeout // the timeout this convergence used
   *   stats.runs    // total times this convergence ran an assertion
   *   stats.value   // last returned value from the queue
   *   stats.queue   // array of other stats for each assertion
   * })
   * ```
   *
   * @returns {Promise}
   */
  run() {
    let start = Date.now();

    let stats = {
      start,
      runs: 0,
      end: start,
      elapsed: 0,
      value: undefined,
      timeout: this._timeout,
      queue: []
    };

    // reduce to a single promise that runs each item in the queue
    return this._queue.reduce((promise, subject, i) => {
      // the last subject will receive the remaining timeout
      if (i === (this._queue.length - 1)) {
        subject = Object.assign({ last: true }, subject);
      }

      return promise.then(ret => {
        if (subject.assertion) {
          return runAssertion.call(this, subject, ret, stats);
        } else if (subject.callback) {
          return runCallback.call(this, subject, ret, stats);
        }
      });
    }, Promise.resolve())
      // always resolve with the stats object
      .then(() => stats);
  }

  /**
   * By being thennable we can enable the usage of async/await syntax
   * with convergences. This allows us to naturally chain convergences
   * without calling `#run()`.
   *
   * ``` javascript
   * async function click(selector) {
   *   // will resolve when the element exists and gets clicked
   *   await new Convergence()
   *     .when(() => {
   *       let $node = document.querySelector('.element')
   *       return !!$node && $node
   *     })
   *     .do(($node) => {
   *       $node.click()
   *     })
   * }
   * ```
   *
   * The convergence thennable method immediately invokes `#run()` and
   * resolves with the last returned value from the convergence's
   * queue. This allows us to await for values from a convergence.
   *
   * ``` javascript
   * let find = (selector) => new Convergence().when(() => {
   *   let $node = document.querySelector('.element')
   *   return !!$node && $node
   * })
   *
   * async function fill(selector, value) {
   *   let $node = await find(selector)
   *   $node.value = value
   * }
   * ```
   *
   * @private
   * @returns {Promise}
   */
  then() {
    // resolve with the value of the last function in the queue
    let promise = this.run().then(({ value }) => value);
    // pass promise arguments onward
    return promise.then.apply(promise, arguments);
  }
}

// static `isConvergence` method
Convergence.isConvergence = isConvergence;

export default Convergence;
